### PR TITLE
[WIP]Add coverlet smoke test

### LINKED ISSUE #2297: Add smoke test for coverlet integration
We should add a smoke test for Coverlet integration, to allow safe update to the embdeded integration code.

cc: @vagisha-nidhi @AbhitejJohn @singhsarab

### PR DESCRIPTION
**Pull request by MarcoRossignoli**
closes https://github.com/microsoft/vstest/issues/2297

cc: @AbhitejJohn  @nohwnd @jakubch1

### CODE REVIEW

**Comment by MarcoRossignoli** [comment]
@nohwnd @jakubch1 I have strange failure running smoke test on CI and local with script
```
test.cmd -f netcoreapp2.1 -p smoke -filter coverlet
...
... .. .. .. .StdOut:
IntegrationTestBase.Execute: Starting dotnet.exe
IntegrationTestBase.Execute: Path = D:\git\vstest\artifacts\testArtifacts\dotnet\dotnet.exe
IntegrationTestBase.Execute: Arguments = test --collect:"XPlat Code Coverage" D:\git\vstest\test\TestAssets\CoverletCoverageTestProject\CoverletCoverageTestProject.csproj --diag:coverletcoverage.a41eac5e1afb450188c78916d35d293e.log
IntegrationTestBase.Execute: Total execution time: 00:00:02.2606459
IntegrationTestBase.Execute: stdError =

IntegrationTestBase.Execute: stdOut = D:\git\vstest\artifacts\testArtifacts\dotnet\sdk\3.1.100\Microsoft.TestPlatform.targets(61,5): error MSB4018: The "Microsoft.TestPlatform.Build.Tasks.VSTestLogsTask" task failed unexpectedly. [D:\git\vstest\test\TestAssets\CoverletCoverageTestProject\CoverletCoverageTestProject.csproj]
D:\git\vstest\artifacts\testArtifacts\dotnet\sdk\3.1.100\Microsoft.TestPlatform.targets(61,5): error MSB4018: System.IO.FileNotFoundException: Could not load file or assembly 'Microsoft.VisualStudio.CodeCoverage.Shim, Version=15.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified. [D:\git\vstest\test\TestAssets\CoverletCoverageTestProject\CoverletCoverageTestProject.csproj]
D:\git\vstest\artifacts\testArtifacts\dotnet\sdk\3.1.100\Microsoft.TestPlatform.targets(61,5): error MSB4018: File name: 'Microsoft.VisualStudio.CodeCoverage.Shim, Version=15.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' [D:\git\vstest\test\TestAssets\CoverletCoverageTestProject\CoverletCoverageTestProject.csproj]
D:\git\vstest\artifacts\testArtifacts\dotnet\sdk\3.1.100\Microsoft.TestPlatform.targets(61,5): error MSB4018:    at Microsoft.TestPlatform.Build.Tasks.VSTestLogsTask.set_LogType(String value) [D:\git\vstest\test\TestAssets\CoverletCoverageTestProject\CoverletCoverageTestProject.csproj]
D:\git\vstest\artifacts\testArtifacts\dotnet\sdk\3.1.100\Microsoft.TestPlatform.targets(61,5): error MSB4018:  [D:\git\vstest\test\TestAssets\CoverletCoverageTestProject\CoverletCoverageTestProject.csproj]    D:\git\vstest\artifacts\testArtifacts\dotnet\sdk\3.1.100\Microsoft.TestPlatform.targets(61,5): error MSB4018:  [D:\git\vstest\test\TestAssets\CoverletCoverageTestProject\CoverletCoverageTestProject.csproj]    D:\git\vstest\artifacts\testArtifacts\dotnet\sdk\3.1.100\Microsoft.TestPlatform.targets(61,56): error MSB4026: The "LogType=BuildStarted" parameter for the "Microsoft.TestPlatform.Build.Tasks.VSTestLogsTask" task is invalid. [D:\git\vstest\test\TestAssets\CoverletCoverageTestProject\CoverletCoverageTestProject.csproj]
D:\git\vstest\artifacts\testArtifacts\dotnet\sdk\3.1.100\Microsoft.TestPlatform.targets(61,5): error MSB4063: The "Microsoft.TestPlatform.Build.Tasks.VSTestLogsTask" task could not be initialized with its input parameters.  [D:\git\vstest\test\TestAssets\CoverletCoverageTestProject\CoverletCoverageTestProject.csproj]
```
It works inside VS and also if I run command by hand on console
```
D:\git\vstest\artifacts\testArtifacts\dotnet\dotnet.exe test --collect:"XPlat Code Coverage" D:\git\vstest\test\TestAssets\CoverletCoverageTestProject\CoverletCoverageTestProject.csproj --diag:coverletcoverage.a41eac5e1afb450188c78916d35d293e.log
...
Starting test execution, please wait...
Logging Vstest Diagnostics in file: D:\git\vstest\coverletcoverage.a41eac5e1afb450188c78916d35d293e.log

A total of 1 test files matched the specified pattern.

Attachments:
  D:\git\vstest\test\TestAssets\CoverletCoverageTestProject\TestResults\45cdee0f-740f-40b4-8fd2-db9f7741f4ad\coverage.cobertura.xml
Test Run Successful.
Total tests: 1
     Passed: 1
 Total time: 1,5941 Seconds
```
Is it expected, am I missing something?

File: test/TestAssets/CoverletCoverageTestProject/CoverletCoverageTestProject.csproj [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  <Import Project="..\..\..\scripts\build\TestPlatform.Dependencies.props" />
+  <!-- Package dependency versions -->
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp2.1</TargetFramework>
+    <IsPackable>false</IsPackable>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.5.0" />
```
Comment (nohwnd) [inline test/TestAssets/CoverletCoverageTestProject/CoverletCoverageTestProject.csproj]: This should use the variables from the TestPlatform.Dependencies to keep the versions up to date. I am currently making changes to run the acceptance tests against the version that is built in the repo, not against the previous version. We will need to rebase this on top of the other change to allow this test to pass, othewise it would use a way old version of the dependencies.

File: test/Microsoft.TestPlatform.SmokeTests/DataCollectorTests.Coverlet.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    using System.Linq;
+    using System.Reflection;
+
+    [TestClass]
+    public class DataCollectorTestsCoverlets : IntegrationTestBase
+    {
+        [TestMethod]
+        public void RunCoverletCoverage()
+        {
+            string projectFileName = this.GetProjectAssetFullPath("CoverletCoverageTestProject.csproj", "CoverletCoverageTestProject.csproj");
```
Comment (nohwnd) [inline test/Microsoft.TestPlatform.SmokeTests/DataCollectorTests.Coverlet.cs]: I am confused why is this doubled, please comment.

File: test/Microsoft.TestPlatform.SmokeTests/DataCollectorTests.Coverlet.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+            string logId = Guid.NewGuid().ToString("N");
+            this.InvokeDotnetTest($"--collect:\"XPlat Code Coverage\" {projectFileName} --diag:coverletcoverage.{logId}.log");
+            var currentDir = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
+
+            // Verify vstest.console.dll CollectArgumentProcessor fix codeBase for coverlet package
```
Comment (nohwnd) [inline test/Microsoft.TestPlatform.SmokeTests/DataCollectorTests.Coverlet.cs]: Not sure what this comment means, could you rephrase it to be more concrete, or explain to me what it means in a comment here?

File: scripts/build.ps1 [hunk 4]
```diff
@@ -228,6 +228,10 @@ function Publish-PatchedDotnet {
     Write-Log "Publish-PatchedDotnet: Copy VSTest task artifacts to local dotnet installation to allow `dotnet test` to run with it"
     $buildArtifactsPath = "$env:TP_ROOT_DIR\src\Microsoft.TestPlatform.Build\bin\$TPB_Configuration\$TPB_TargetFrameworkNS2_0\*"
     Copy-Item $buildArtifactsPath $dotnetTestArtifactsSdkPath -Force
+
+    Write-Log "Publish-PatchedDotnet: Copy vstest.console.dll artifact to local dotnet installation to allow `dotnet test` to run with it"
```
Comment (nohwnd) [inline scripts/build.ps1]: During my current work on the acceptance tests I found that this is falling short, because I am patching the task definition and the build but not the console and testhost. Would this be able to run via the InvokeVstest (or the other alike methods that run vsconsole but not dotnet test)? I tried taking the contents of the package that we ultimately insert into `dotnet test` (the CLI package), and it broke the runner, and I would rather avoid copying "random" dlls into the local copy of dotnet because then we can't really be sure about what we are testing against and if it reflects the real world.

File: test/Microsoft.TestPlatform.SmokeTests/DataCollectorTests.Coverlet.cs (hunk 3 above)
Comment (MarcoRossignoli) [reply test/Microsoft.TestPlatform.SmokeTests/DataCollectorTests.Coverlet.cs]: With this PR https://github.com/microsoft/vstest/pull/2288 we removed custom code to allow coverlet lib to load in-proc collector.
That PR added some code inside `CollectArgumentProcessor` where we looking for "base path" of `coverlet.collector.dll` expected inside nuget package default folder.
When we find the lib we'll emit a log https://github.com/microsoft/vstest/pull/2288/files#diff-6f2506b5205b023825192043bd33a256R138
This assert check that we're sure that we've updated collector setting code base with full path, otherwise without "custom coverlet code" inside `ProxyExecutionManager` coverlet dll won't be resolved inside testhost.

File: test/Microsoft.TestPlatform.SmokeTests/DataCollectorTests.Coverlet.cs (hunk 2 above)
Comment (MarcoRossignoli) [reply test/Microsoft.TestPlatform.SmokeTests/DataCollectorTests.Coverlet.cs]: It's a tmp typo...this PR is WIP I've updated on my local branch with `string projectFileName = this.GetProjectFullPath("CoverletCoverageTestProject.csproj");` but I'll wait to push until test work...I need an help to undestand if my issue is something already saw https://github.com/microsoft/vstest/pull/2330#issuecomment-586622410

File: scripts/build.ps1 (hunk 4 above)
Comment (MarcoRossignoli) [reply scripts/build.ps1]: I cannot use VsTest...because the base path of coverlet directory(nuget package) is injected through msbuild prop https://github.com/tonerdo/coverlet/blob/master/src/coverlet.collector/build/netstandard1.0/coverlet.collector.targets
I should pass `/testadapterpath:` with path https://github.com/microsoft/vstest/issues/2278#issuecomment-586640772 but it depends on where we run tests and on which plat(this is a less important issue, coverlet pack at the moment has got only one tfm netstandard1.0)

File: test/TestAssets/CoverletCoverageTestProject/CoverletCoverageTestProject.csproj (hunk 1 above)
Comment (MarcoRossignoli) [reply test/TestAssets/CoverletCoverageTestProject/CoverletCoverageTestProject.csproj]: Ok I'll rebase when update will be in place...or do you want that I create a new prop `NETTestSdkBuildVersion` and for now fix the version and you'll update with var?

Comment (nohwnd) [reply test/TestAssets/CoverletCoverageTestProject/CoverletCoverageTestProject.csproj]: That's not necessary I would have to go and remove that variable afterwards. I am fighting with the acceptance tests to pass on the server, but hopefully it won't take more than a day more to get it to pass... 🤞 (or at least I hope because it is boring and exhausting 😁 )

File: scripts/build.ps1 (hunk 4 above)
Comment (nohwnd) [reply scripts/build.ps1]: Let's figure that out later, imho finding the adapter should not be that hard, there will be some tests already doing it. But I would definitely love to get the dotnet test working with the currently built taks + console + testhost, because testing dotnet test more extensively would be nice. Maybe I was just doing something wrong.

File: test/TestAssets/CoverletCoverageTestProject/CoverletCoverageTestProject.csproj (hunk 1 above)
Comment (MarcoRossignoli) [reply test/TestAssets/CoverletCoverageTestProject/CoverletCoverageTestProject.csproj]: 😆 🤞

File: scripts/build.ps1 (hunk 4 above)
Comment (MarcoRossignoli) [reply scripts/build.ps1]: I'll take a look if `InvokeVstest`  is used by other tests.

File: test/TestAssets/CoverletCoverageTestProject/CoverletCoverageTestProject.csproj (hunk 1 above)
Comment (nohwnd) [reply test/TestAssets/CoverletCoverageTestProject/CoverletCoverageTestProject.csproj]: #2340 is now in master

Comment (MarcoRossignoli) [reply test/TestAssets/CoverletCoverageTestProject/CoverletCoverageTestProject.csproj]: Great!Any news on strange error https://github.com/microsoft/vstest/pull/2330#issuecomment-586622410?

**Comment by MarcoRossignoli** [comment]
https://github.com/microsoft/vstest/pull/2340 was pretty invasive and changed project guid...I don't want to lose too much time on fixing merging issue, I'll open new one with new test.
